### PR TITLE
Remove API export decorator from base classes

### LINKED ISSUE #823: Remove API export decorator from base classes
We should remove the API export decorator from base classes like [`Backbone`](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/backbone.py#L26-L27), [`Preprocessor`](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/preprocessor.py#L22-L23), [`Task`](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/task.py#L26-L27) and [`Tokenizer`](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/tokenizers/tokenizer.py#L22-L23). These classes are meant to be used internally.

### PR DESCRIPTION
**Pull request by shivance**
Fixes #823 

Needs @mattdangerw 's confirmation, keeping [this comment](https://github.com/keras-team/keras-nlp/pull/814#discussion_r1131716251) in mind.

### CODE REVIEW

**Comment by mattdangerw** [comment]
Actually Sampler and Tokenizer we do want to export right now, as we want library users to be able to write subclasses for these.

This looks good for Backbone, Task, and Preprocessor!

File: keras_nlp/models/backbone.py [hunk 1]
```diff
@@ -18,12 +18,10 @@
 
 from tensorflow import keras
 
-from keras_nlp.api_export import keras_nlp_export
 from keras_nlp.utils.python_utils import classproperty
 from keras_nlp.utils.python_utils import format_docstring
 
 
-@keras_nlp_export("keras_nlp.models.Backbone")
```
Comment (abheesht17) [inline keras_nlp/models/backbone.py:26]: Shouldn't we replace this with 

```
@keras.utils.register_keras_serializable(package="keras_nlp")
```

? And elsewhere?

Comment (mattdangerw) [reply keras_nlp/models/backbone.py:26]: Good call.

So in total:
 - Sampler, Tokenizer: leave as is
 - Task, Backbone, Preprocessor: do not export, but do register as serializable 

Thanks!

**Comment by shivance** [comment]
Hey @mattdangerw, changes us are up, and PR is RTM !
btw could you quickly summarize how `keras_nlp_export` differs from `register_keras_serializable` in terms of functioning, (apart from export logic).
Thanks !

**Comment by mattdangerw** [comment]
Sure! `keras_nlp_export` just does `register_keras_serializable` under the hood. https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/api_export.py

Generally it is good practice to make sure every object in the library is registered as serializable if it (or a parent class) has a `from_config()`. So that is basically all layers, tokenizers, models, metrics, samplers, etc. Whether or not they are part of the API.

`keras_nlp_export` means we are adding this to the API of the library. Everything with an export annotation will be under our API contract and documented on keras.io. We have things (like the WhisperEncoder we just added) that we don't want to export and are essentially private details of our library. This includes `Task` and `Backbone` classes for now. We don't want people writing their own subclasses and filing bugs if we break compatibility on these classes.

So basically all "keras objects" should have either an register call or an export call. Register if its private functionality, export if its public.

**Review by mattdangerw — APPROVED** [review]
LGTM! Will merge as soon as CI is green.
